004-testing.md

```diff
@@ -0,0 +1,50 @@
+
+## Proposal
+
+Well, we should write tests, of course! Here's our current "best practices" for testing:
+
+### Philosophy
+There are lots of opinions on how to write tests! We try be pragmatic.
+We write tests to save ourselves from doing tedious QA work, and our tests reflect that: they're the steps that an engineer or PM would use to test a given function, component, API endpoint, or user flow... but translated into a set of instructions so a machine can do it for us!
+
+We [don't practice strict TDD](https://dhh.dk/2014/tdd-is-dead-long-live-testing.html) or pursue 100% test coverage. Instead, we try to get the most "bang for our buck" by writing [fewer higher-level integration tests](https://youtu.be/Fha2bVoC8SE) that test the real functionality that our application exposes, rather than isolated pieces.
+
+In [Phoenix](https://github.com/DoSomething/phoenix), this means the majority of our test coverage comes from [Cypress](https://www.cypress.io/) browser tests, which step through full application flows just like a real user. We also maintain some unit tests, written with [Jest](https://jestjs.io/) & [Enzyme](https://airbnb.io/enzyme/), for more complicated components. In backend Laravel applications, like [Northstar](https://github.com/DoSomething/northstar) and [Rogue](https://github.com/DoSomething/rogue), the majority of our test coverage comes from [PHPUnit](https://laravel.com/docs/5.5/http-tests) functional tests, which simulate real API requests.
+
+### Test Structure
+
+In general, all of our tests follow the same basic structure: "given this scenario, when I do X, I see Y as a result". For example,  ["given this JSON payload, the ContentfulEntry renders the appropriate component"](https://github.com/DoSomething/phoenix-next/blob/dc9713b568060a3a459c715ffb97854321d4ae10/resources/assets/components/ContentfulEntry/ContentfulEntry.test.js#L19-L26) or ["given a user isn't signed up for the campaign, if they click on the Join Now button, then they should see the affirmation modal"](https://github.com/DoSomething/phoenix-next/blob/dc9713b568060a3a459c715ffb97854321d4ae10/cypress/integration/campaign-signup.js#L35-L53).
+
+```js
+describe('Test Suite name', () => {
+   it('should do the things', () => {
+      // 1. Given some state, such as from a mock or factory...
+      // 2. When something happens (click, HTTP request, function call...)
+      // 3. Assert the result matches our expectations!
+   }
+});
+```
+
+### Process
+We like to write tests while adding features or fixing bugs, because writing tests help us "check our work" & demonstrate functionality in pull requests, and because otherwise it's easy to forget! We're still catching up on tests we skipped writing when we originally built Phoenix, Chompy, and other applications.
```

Comment:
It might be nice for Vanessa and I to try and bring some existing applications up to speed with testing in order to get more familiar with them! Depending on the focus of Q4, that might be a nice way for us to ramp up.

Reply:
Yesss! I think this is a wonderful idea. 😸